system/jlib/jfile.cpp

```diff
@@ -7484,3 +7484,95 @@ IAPICopyClient * createApiCopyClient(IStorageApiInfo * source, IStorageApiInfo *
     }
     return nullptr;
 }
+
+class CBlockFileIO : public CInterfaceOf<IFileIO>
+{
+private:
+    Owned <IFileIO> io;
+    offset_t fileSize = 0;
+    static size32_t blockSize;
+    static thread_local byte *buffer;
+    static thread_local offset_t lastReadPos;
+    static thread_local size32_t lastReadLen;
+    static thread_local CBlockFileIO *lastReadCaller;
+    static bool onThreadTerm(bool isPooled);
+public:
+    CBlockFileIO(IFileIO *_io) : io(_io), fileSize(_io->size()) {}
+    ~CBlockFileIO();
+    virtual size32_t read(offset_t pos, size32_t len, void * data) override;
+    virtual offset_t size() override { return fileSize; }
+    virtual size32_t write(offset_t pos, size32_t len, const void * data) override { throwUnexpected(); }
+    virtual offset_t appendFile(IFile *file,offset_t pos=0,offset_t len=(offset_t)-1) override { throwUnexpected(); }
+    virtual void setSize(offset_t size) override { throwUnexpected(); }
+    virtual void flush() override { throwUnexpected(); }
+    virtual void close() override { io->close(); }
+    virtual unsigned __int64 getStatistic(StatisticKind kind) override { return io->getStatistic(kind); }
+
+    inline static void setBlockSize(size32_t _blockSize) { blockSize = _blockSize; }
+};
+
+size32_t CBlockFileIO::blockSize = 64*1024;
```

Comment:
Might be better if these were static variables rather than static members of the class (different visibility and possibly less efficient access.).  May be identical though..